File: src/converter/plugins/convertTypePredicate.ts
import ts from "typescript";
import {createSimplePlugin} from "../plugin.js";
import {CheckCoverageService, checkCoverageServiceKey} from "./CheckCoveragePlugin.js";
import {TypeScriptService, typeScriptServiceKey} from "./TypeScriptPlugin.js";

export const convertTypePredicate = createSimplePlugin((node, context, render) => {
    if (!ts.isTypePredicateNode(node)) return null

    const checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)
    checkCoverageService?.deepCover(node)

    const typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)

    // TODO: support contracts

    const type = Boolean(node.assertsModifier)
        ? "Unit"
        : "Boolean"

    return `${type} /* ${typeScriptService?.printNode(node)} */`
})

File: gradle-plugin/karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/convertTypePredicate.kt
package io.github.sgrishchenko.karakum.extension.plugins

import io.github.sgrishchenko.karakum.extension.createSimplePlugin
import typescript.Node
import typescript.isTypePredicateNode

val convertTypePredicate = createSimplePlugin plugin@{ node: Node, context, _ ->
    if (!isTypePredicateNode(node)) return@plugin null

    val checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)
    checkCoverageService?.deepCover(node)

    val typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)

    // TODO: support contracts

    val type = if (node.assertsModifier != null) "Unit" else "Boolean"

    "$type /* ${typeScriptService?.printNode(node)} */"
}
